Moving flag as deleted behavior to a redis set.

ryansch opened the pull request:
This is a large refactor combined with new functionality.
I've removed the flag as deleted job and rolled that functionality into
the current delta job.  When a model is indexed I create the delta job
and add the flag as deleted document id to a redis set.  The delta job
handles getting the set of document ids to flag as deleted without
racing.

This should also allow multiple resque workers to work off different
delta index jobs (on distinct indices) concurrently.

ryansch committed:
Moving flag as deleted behavior to a redis set.

This is a large refactor combined with new functionality.
I've removed the flag as deleted job and rolled that functionality into
the current delta job.  When a model is indexed I create the delta job
and add the flag as deleted document id to a redis set.  The delta job
handles getting the set of document ids to flag as deleted without
racing.

This should also allow multiple resque workers to work off different
delta index jobs (on distinct indices) concurrently.

ryansch commented:
I'll be testing this in production.  I'll let you know how it goes.

ryansch committed:
Updating mock_redis.

ryansch commented:
We'll also need to decide if we want to apply the same approach to flying_sphinx.

ryansch commented:
I've found a bug with validating flag as deleted ids this way:

``` ruby
    unless flag_as_deleted_ids.empty?
      # Filter out the ids that aren't present in sphinx
      flag_as_deleted_ids = ThinkingSphinx::Search.bundle_searches(flag_as_deleted_ids) do |sphinx, id|
        sphinx.search_for_ids([], :index => index, :id_range => id..id)
      end.map(&:to_a).flatten

      unless flag_as_deleted_ids.empty?
        # Each hash element should be of the form { id => [1] }
        flag_hash = Hash[*flag_as_deleted_ids.collect {|id| [id, [1]] }.flatten(1)]

        config.client.update(index, ['sphinx_deleted'], flag_hash)
      end
    end
```

As you can see, I'm batching the queries together with the multi-query stuff but I've run into the 32 query limit.  I don't think it's practical to ask admins to change a sphinx config value just for this addon so I'm not sure which of the following approaches I should use.
1. Batch the batches - Partition my ids up into groups of 32 and send each partition to sphinx with `bundle_searches`.
2. Persistent connection - Figure out how to `client.open` and `client.close` and just call `ThinkingSphinx.search_for_id` in a loop between them.
3. Combine both - Set up a persistent connection AND batch the batches so all batches go over the same connection.

Number 3 seems like the best approach but I've found people with issues with persistent connections and sphinx via google.

@freelancing-god: Could you share your thoughts here?

ryansch commented:
Update:  In researching this problem I've realized that I'm accidentally converting sphinx document ids to sphinx internal ids (usually the ActiveRecord model unique id).

I can see in the docs that sphinx exposes the document id as `@id` when sorting.  I'll check to see if that works for a query.
If not, a better approach would be to patch thinking-sphinx itself to add the sphinx document id as an attribute I can search on.

pat commented:
Agreed, option 3 is best, but may be tricky. At the very least, go with option 1. As for searching on specific document ids, this is exposed through the :id_range option:
https://github.com/freelancing-god/thinking-sphinx/blob/master/lib/thinking_sphinx/search_methods.rb#L397-399

If you're dealing with sets of values that aren't ranges, then that gets trickier. Maybe Sphinx allows filtering on @id, but I'm not sure. You should be able to sort by @id, though.

ryansch commented:
:id_range has to be contiguous and I can't guarantee that the set of ids I want to work on is contiguous.

ryansch committed:
Adding a verbose switch for smart index.

ryansch committed:
Fixing flag as deleted id filtering

morganick commented:
We just pushed this into production... and is so hot. We have 3 workers working on deltas now. : ) GJ @ryansch

ryansch commented:
:shipit:

ryansch committed:
Batching id check into batches of 4096

Sphinx doesn't like filtering on attribute lists longer than 4096.

ryansch committed:
Actually use the persistent connection.

ryansch committed:
Removing persistent sphinx connection

ryansch commented:
It turns out that I couldn't get the persistent connections to sphinx to work.  It would actually crash my resque workers.  Here's what I was trying to do:

``` ruby
config = ThinkingSphinx::Configuration.instance
client = config.client
client.open

<some number>.times do
  ThinkingSphinx.search_for_ids(foo, :client => client)
  client.reset
end

client.close
```

agibralter committed:
Allow test suite to run on ruby 1.9.x.

agibralter committed:
Make sure to wait for sphinx to catch up.

I found that a couple scenarios were failing occasionally depending on
my computer's speed.

agibralter commented:
Could you change the pull request to be for your branch `release`? Also, since this is such a big merge, I'm going to open a pull request from my own `release` branch against your `release` branch to resolve minor issues in this pull request. Once we close/resolve my pull request against your fork, and you merge the commits that make sense from my pull request, we can then finally merge this pull. Nice and complicated? :)

ryansch commented:
The only difference between this branch and my release branch is that I've changed the info in the gemspec to fork it so I can easily deploy it to production.

Would you instead like me to rebase this down to a single commit?

ryansch commented:
I'm fine if you want to open pull requests against this branch in the same manner that you did in ryansch/ts-resque-delta#2.

agibralter commented:
Oh no, this merge request is fine. No need to rebase.

The only reason I asked you to change the commit range is so that we can also merge any changes from my pull request: https://github.com/ryansch/ts-resque-delta/pull/2

But Crap. I thought you could change the commit range of a pull request... Ok I'll open a new pull request against combined-job.

ryansch committed:
Moved test opts to Rakefile and cucumber.yml

ryansch committed:
Use rspec's array matcher

agibralter commented:
Totally understand what you mean now! We just got hammered by a big article: http://andrewsullivan.thedailybeast.com/2011/12/dish-readers-who-are-you.html

http://cl.ly/2Y43182j0y0q24403T2a -- The ts_delta queue is just getting hammered by FlagAsDeletedJobs. Alright... so I'm down to pull in your changes if you think everything is stable. What do you think about that one `Scenario` we were talking about? The one about the smart indexer clearing the delta queue?

ryansch commented:
I've verified that it tests what it's supposed to so I think we should keep it as is.  We're still seeing some EAGAIN and DirtyExit errors but our indexes are staying up to date.  I still haven't had time to chase those down partly because it's working anyway and partly because my parents are going to be here in just a few hours to celebrate Christmas with us.  :-)

morganick commented:
@agibralter I can concur that everything is running smoothly other than those two errors that get thrown. That has had no impact on our ability to index.  We normally would have had 200k-300k in the delta by mid day. We now hover around 100 : ) with three workers going. Note there will be elevated load on the db because you will be indexing so much faster. Gratz on the article :)

agibralter commented:
Thank you! :) Ok merged. Will push a gem soon.

agibralter commented on lib/thinking_sphinx/deltas/resque_delta/core_index.rb:
```diff
@@ -0,0 +1,101 @@
+  #
+  # Returns nothing.
+  def lock_deltas
+    sphinx_indices.each { |index_name| lock_delta(index_name) }
```
This method is showing up as undefined. Have you not run into that problem?

agibralter replied:
Only a problem if you try and run the normal indexer...

ryansch replied:
I guess I should have tested that!

On Dec 21, 2011, at 11:21, Aaron Gibralter
reply@reply.github.com
wrote:

> > -  # index_name - The String index prefix.
> > -  #
> > -  # Examples
> > -  #
> > -  #   unlock_delta('foo')
> > -  #
> > -  # Returns nothing.
> > -  def unlock_delta(index_name)
> > -    ThinkingSphinx::Deltas::ResqueDelta.unlock("#{index_name}_delta")
> > -  end
> >   +
> > -  # Public: Lock all delta indexes against indexing or new index jobs.
> > -  #
> > -  # Returns nothing.
> > -  def lock_deltas
> > -    sphinx_indices.each { |index_name| lock_delta(index_name) }
> 
> Only a problem if you try and run the normal indexer...
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12/files#r309731

agibralter replied:
Heh, I just copied the code from the old rake task and it seems to work. In
a hot fix situation so no time to test! :)

On Wed, Dec 21, 2011 at 2:40 PM, Ryan Schlesinger <
reply@reply.github.com

> wrote:
> 
> > -  # index_name - The String index prefix.
> > -  #
> > -  # Examples
> > -  #
> > -  #   unlock_delta('foo')
> > -  #
> > -  # Returns nothing.
> > -  def unlock_delta(index_name)
> > -    ThinkingSphinx::Deltas::ResqueDelta.unlock("#{index_name}_delta")
> > -  end
> >   +
> > -  # Public: Lock all delta indexes against indexing or new index jobs.
> > -  #
> > -  # Returns nothing.
> > -  def lock_deltas
> > -    sphinx_indices.each { |index_name| lock_delta(index_name) }
> 
> I guess I should have tested that!
> 
> On Dec 21, 2011, at 11:21, Aaron Gibralter
> reply@reply.github.com
> wrote:
> 
> > > -  # index_name - The String index prefix.
> > > -  #
> > > -  # Examples
> > > -  #
> > > -  #   unlock_delta('foo')
> > > -  #
> > > -  # Returns nothing.
> > > -  def unlock_delta(index_name)
> > > -    ThinkingSphinx::Deltas::ResqueDelta.unlock("#{index_name}_delta")
> > > -  end
> > >   +
> > > -  # Public: Lock all delta indexes against indexing or new index jobs.
> > > -  #
> > > -  # Returns nothing.
> > > -  def lock_deltas
> > > -    sphinx_indices.each { |index_name| lock_delta(index_name) }
> > 
> > Only a problem if you try and run the normal indexer...
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/agibralter/ts-resque-delta/pull/12/files#r309731
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12/files#r309786

agibralter commented:
Also... have you guys run into: `NoMethodError: undefined method 'index' for nil:NilClass`

https://gist.github.com/0e1b68863d4c6f46f9a2

ryansch commented:
No, that's a new one...

On Dec 21, 2011, at 13:17, Aaron Gibralter wrote:

> Also... have you guys run into: `NoMethodError: undefined method`index' for nil:NilClass`
> 
> https://gist.github.com/0e1b68863d4c6f46f9a2
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239738

agibralter commented:
Hmm I wonder if it's because I'm using it with thinking-sphinx 1.4.10...
could that be it?

On Wed, Dec 21, 2011 at 4:19 PM, Ryan Schlesinger <
reply@reply.github.com

> wrote:
> 
> No, that's a new one...
> 
> On Dec 21, 2011, at 13:17, Aaron Gibralter wrote:
> 
> > Also... have you guys run into: `NoMethodError: undefined method`index'
> > for nil:NilClass`
> > 
> > https://gist.github.com/0e1b68863d4c6f46f9a2
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > 
> > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239738
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239756

ryansch commented:
We're running with 1.4.10 in production.

On Dec 21, 2011, at 13:23, Aaron Gibralter wrote:

> Hmm I wonder if it's because I'm using it with thinking-sphinx 1.4.10...
> could that be it?
> 
> On Wed, Dec 21, 2011 at 4:19 PM, Ryan Schlesinger <
> reply@reply.github.com
> 
> > wrote:
> > 
> > No, that's a new one...
> > 
> > On Dec 21, 2011, at 13:17, Aaron Gibralter wrote:
> > 
> > > Also... have you guys run into: `NoMethodError: undefined method`index'
> > > for nil:NilClass`
> > > 
> > > https://gist.github.com/0e1b68863d4c6f46f9a2
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > 
> > > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239738
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239756
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3239808

agibralter commented:
Ah, you know what -- it may be that searchd was running when I kind of hot patched this in. Maybe searchd was missing some ids from left over delta jobs... who knows... Hopefully when I get around to this bigger deploy with TS 2.0 and Rails 3.1 in a few days, these errors will go away.

agibralter commented:
Hmm, also, what version of sphinx are you guys running?

ryansch commented:
I wonder if it would be worth using appraisal or Travis to test
against bundled rails 2 and 3 apps.

On Dec 21, 2011, at 13:51, Aaron Gibralter
reply@reply.github.com
wrote:

> Ah, you know what -- it may be that searchd was running when I kind of hot patched this in. Maybe searchd was missing some ids from left over delta jobs... who knows... Hopefully when I get around to this bigger deploy with TS 2.0 and Rails 3.1 in a few days, these errors will go away.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3240161

ryansch commented:
0.9.9 in production and 2.0.2 locally.  I'll be upgrading to 2.0.2 soon.

On Dec 21, 2011, at 14:43, Aaron Gibralter
reply@reply.github.com
wrote:

> Hmm, also, what version of sphinx are you guys running?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3240805

agibralter commented:
So your stack is TS 1.4.10 + Rails 2.3.x?

On Thu, Dec 22, 2011 at 4:10 PM, Ryan Schlesinger <
reply@reply.github.com

> wrote:
> 
> I wonder if it would be worth using appraisal or Travis to test
> against bundled rails 2 and 3 apps.
> 
> On Dec 21, 2011, at 13:51, Aaron Gibralter
> reply@reply.github.com
> wrote:
> 
> > Ah, you know what -- it may be that searchd was running when I kind of
> > hot patched this in. Maybe searchd was missing some ids from left over
> > delta jobs... who knows... Hopefully when I get around to this bigger
> > deploy with TS 2.0 and Rails 3.1 in a few days, these errors will go away.
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > 
> > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3240161
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3254904

ryansch commented:
Yeah.

We're working on upgrading to rails 3.1 but we're not running that in production yet.

On Dec 22, 2011, at 13:26, Aaron Gibralter wrote:

> So your stack is TS 1.4.10 + Rails 2.3.x?
> 
> On Thu, Dec 22, 2011 at 4:10 PM, Ryan Schlesinger <
> reply@reply.github.com
> 
> > wrote:
> > 
> > I wonder if it would be worth using appraisal or Travis to test
> > against bundled rails 2 and 3 apps.
> > 
> > On Dec 21, 2011, at 13:51, Aaron Gibralter
> > reply@reply.github.com
> > wrote:
> > 
> > > Ah, you know what -- it may be that searchd was running when I kind of
> > > hot patched this in. Maybe searchd was missing some ids from left over
> > > delta jobs... who knows... Hopefully when I get around to this bigger
> > > deploy with TS 2.0 and Rails 3.1 in a few days, these errors will go away.
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > 
> > > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3240161
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3254904
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3255062

agibralter commented:
Ah, we just went through that hell. About to finally be done with it...

On Thu, Dec 22, 2011 at 6:23 PM, Ryan Schlesinger <
reply@reply.github.com

> wrote:
> 
> Yeah.
> 
> We're working on upgrading to rails 3.1 but we're not running that in
> production yet.
> 
> On Dec 22, 2011, at 13:26, Aaron Gibralter wrote:
> 
> > So your stack is TS 1.4.10 + Rails 2.3.x?
> > 
> > On Thu, Dec 22, 2011 at 4:10 PM, Ryan Schlesinger <
> > reply@reply.github.com
> > 
> > > wrote:
> > > 
> > > I wonder if it would be worth using appraisal or Travis to test
> > > against bundled rails 2 and 3 apps.
> > > 
> > > On Dec 21, 2011, at 13:51, Aaron Gibralter
> > > reply@reply.github.com
> > > wrote:
> > > 
> > > > Ah, you know what -- it may be that searchd was running when I kind of
> > > > hot patched this in. Maybe searchd was missing some ids from left over
> > > > delta jobs... who knows... Hopefully when I get around to this bigger
> > > > deploy with TS 2.0 and Rails 3.1 in a few days, these errors will go
> > > > away.
> > > > 
> > > > ---
> > > > 
> > > > Reply to this email directly or view it on GitHub:
> > > 
> > > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3240161
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > 
> > > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3254904
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > 
> > https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3255062
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/12#issuecomment-3256250

agibralter commented:
Uggg, 4000 Errno::EAGAIN: Resource temporarily unavailable - Timeout reading from the socket and 46,000 NoMethodError: undefined method `index' for nil:NilClass yesterday...

Hmm, as for the first: http://corner.squareup.com/2011/07/distributed_faults.html. What's strange though is I have my timeout set very high in my redis.conf. Also: https://github.com/ezmobius/redis-rb/issues/84 -- seems like redis-rb doesn't handle the case natively. Maybe it makes sense to just wrap the redis calls in blocks that rescue from EAGAIN. Problem is probably this method:

``` ruby
    def get_subset_for_processing(core_name)
      # Copy set to temp
      Resque.redis.sunionstore temp_name(core_name), set_name(core_name)
      # Store (set - temp) into set.  This removes all items we copied into temp from set.
      Resque.redis.sdiffstore set_name(core_name), set_name(core_name), temp_name(core_name)
      # Merge processing and temp together and store into processing.
      Resque.redis.sunionstore processing_name(core_name), processing_name(core_name), temp_name(core_name)

      Resque.redis.del temp_name(core_name)
    end
```

Should it get wrapped in a multi?

As for the NoMethodError, I can't figure out where a method `index` is getting called in this backtrace: https://gist.github.com/0e1b68863d4c6f46f9a2 -- I only see `index` passed as an argument... very weird...

agibralter commented:
Ahh ok, this line: https://github.com/freelancing-god/riddle/blob/v1.5.0/lib/riddle/client.rb#L568

agibralter commented:
Ah, found it! It's related to this shuffle business: https://github.com/freelancing-god/thinking-sphinx/issues/314#issuecomment-3111739 and the way I pass in the server config variable in sphinx.yml. :)

ryansch commented:
Wow your error rate is quite a bit higher than ours.  I had thought about using redis multi around those commands but I don't need them to be atomic all together for that to work correctly.

Have you been able to fix your NoMethodError now?
